packages/core/src/components/menu/menuItem.tsx

```diff
@@ -174,47 +182,56 @@ export class MenuItem extends AbstractPureComponent<IMenuItemProps, IMenuItemSta
         }
     }
 
-    private liRefHandler = (r: HTMLElement) => (this.liElement = r);
+    private handlePopoverDidOpen = () => {
+        if (this.props.useSmartPositioning) {
+            // Popper.js renders the popover in the DOM before relocating its
+            // position on the next tick. We need to rAF to wait for that to happen.
+            requestAnimationFrame(() => this.maybeAlignSubmenuLeft());
```

Comment:
🤔 Popper.js _should_ be powerful enough to handle all the positioning logic: we should let _it_ decide when to flip to the left, because it likely knows better than us. Happy for this to be a follow-up, though.